circus-train-gcp/src/main/java/com/hotels/bdp/circustrain/gcp/GCPCredentialCopier.java

```diff
@@ -15,77 +15,105 @@
  */
 package com.hotels.bdp.circustrain.gcp;
 
+import static org.apache.commons.lang.StringUtils.isBlank;
+
 import static com.hotels.bdp.circustrain.gcp.GCPConstants.GCP_KEYFILE_CACHED_LOCATION;
 
-import java.io.File;
 import java.io.IOException;
 import java.net.URI;
 import java.net.URISyntaxException;
-import java.util.UUID;
 
-import org.apache.commons.io.FileUtils;
 import org.apache.hadoop.conf.Configuration;
-import org.apache.hadoop.filecache.DistributedCache;
 import org.apache.hadoop.fs.FileSystem;
 import org.apache.hadoop.fs.Path;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import com.hotels.bdp.circustrain.api.CircusTrainException;
+import com.hotels.bdp.circustrain.gcp.context.GCPSecurity;
 
 class GCPCredentialCopier {
   private static final Logger LOG = LoggerFactory.getLogger(GCPCredentialCopier.class);
-  private static final String WORKING_DIRECTORY = System.getProperty("user.dir");
-  private static final String RANDOM_STRING = UUID.randomUUID().toString() + System.currentTimeMillis();
-  private static final String CACHED_CREDENTIAL_NAME = "/ct-gcp-key-" + RANDOM_STRING + ".json";
-  private static final String HDFS_GS_CREDENTIAL_DIRECTORY = "hdfs:///tmp/ct-gcp-" + RANDOM_STRING;
-  private static final String HDFS_GS_CREDENTIAL_ABSOLUTE_PATH = HDFS_GS_CREDENTIAL_DIRECTORY + CACHED_CREDENTIAL_NAME;
 
   private final FileSystem fs;
   private final Configuration conf;
-  private final String credentialProvider;
+  private final GCPSecurity security;
+  private final CredentialProviderRelativePathFactory credentialProviderRelativePathFactory;
+  private final HdfsGsCredentialAbsolutePathFactory hdfsGsCredentialAbsolutePathFactory;
+  private final HdfsGsCredentialDirectoryFactory hdfsGsCredentialDirectoryFactory;
+  private final RandomStringFactory randomStringFactory;
+  private final String credentialsFileRelativePath;
+  private final String randomString;
+  private final Path hdfsGsCredentialDirectory;
+  private final Path hdfsGsCredentialAbsolutePath;
+
+  GCPCredentialCopier(FileSystem fs, Configuration conf, GCPSecurity security) {
+    this(fs, conf, security, new CredentialProviderRelativePathFactory(), new HdfsGsCredentialAbsolutePathFactory(),
+        new HdfsGsCredentialDirectoryFactory(), new RandomStringFactory());
```

Comment:
Remove the `RandomStringFactory` but add it to the constructor of the Factories that need it. 
example: `new CredentialProviderRelativePathFactory(new RandomStringFactory())`